packages/php-wasm/universal/src/lib/php-request-handler.ts

```diff
@@ -339,6 +339,23 @@ export class PHPRequestHandler {
 
 		let fsPath = joinPaths(this.#DOCROOT, normalizedRequestedPath);
 
+		// If we are dealing with a symlink, lets resolve the target
+		// and continue using that instead of the symlink.
+		if (primaryPhp.isSymlink(fsPath)) {
```

Comment:
This would likely be a big change, touching a lot of FSHelper methods, but maybe FSHelper should resolve paths by following symlinks so we don't need this code at all.

Even basic FS operations like [POSIX stat()](https://pubs.opengroup.org/onlinepubs/007904975/functions/stat.html#:~:text=If%20the%20named%20file%20is,in%20the%20%3Csys%2Fstat.) resolve paths containing symlinks:
> If the named file is a symbolic link, the stat() function shall continue pathname resolution using the contents of the symbolic link, and shall return information pertaining to the resulting file if the file exists.

AFAIK, symlinks are normally resolved by the OS / platform libraries, and doing so in FSHelpers would make it so that higher level code like the PHPRequestHandler doesn't have to concern itself with whether or not a given path contains a symlink.

It's more work now, but I think doing this at the FS level would keep the rest of our code simpler and more focused on its actual purpose.

@adamziel, do you have any thoughts on this?

Reply:
That would indeed be a more significant change, but I agree with your point. I'd be happy to help out with that.

Could we consider fixing the FShelpers in a separate PR so in the meantime we have a way to deal with symlinks?

Reply:
> maybe FSHelper should resolve paths by following symlinks so we don't need this code at all.

I'd like to better understand the level of abstraction where the problem is located. Patching FSHelpers would help TypeScript code, but it would do nothing for the PHP code executing via WASM. To help the latter, we'd need to patch the Emscripten Filesystem backends like MEMFS or NODEFS. But do we actually have that problem in PHP?

Reply:
Based on my testing, PHP functions for handling symlinks (such as stat, lstat, readlink, and realpath) appear to function correctly with NODEFS. I haven't verified this for MEMFS and other filesystems, but we could check that.

Having thought about this idea more I think that perhaps it is not ideal to have all helper functions (which primarily are wrappers around the FS API provided by emscripten) to follow symlinks. As I understood it, the idea was to not have to alter the code in the request handler by having methods like `isFile` and `isDir` return true on a symlink that leads to a file or directory respectively. However, as far as i can tell, the FS API is inspired by POSIX filesystem API so things like `stat` and `lstat` should already work. I think that methods like `isFile` and `isDir` should not return `true` if the provided path is a symlink.

As a side note, I encountered issues when attempting to use FS.stat() - it consistently returned undefined in my tests.

Reply:
>  However, as far as i can tell, the FS API is inspired by POSIX filesystem API so things like stat and lstat should already work. I think that methods like isFile and isDir should not return true if the provided path is a symlink.

At the NODEFS and MEMFS level, I think sticking to stat and lstat semantics is the right choice 👍 

At the FSHelpers level, it might make sense to follow symlinks. I've checked a few languages and it seems like their standard library follows symlinks for utilities like isFile and isDir:

[Python](https://docs.python.org/3/library/os.path.html#os.path.isdir):

> os.path.isdir(path)
Return True if path is an [existing](https://docs.python.org/3/library/os.path.html#os.path.exists) directory. This follows symbolic links, so both [islink()](https://docs.python.org/3/library/os.path.html#os.path.islink) and [isdir()](https://docs.python.org/3/library/os.path.html#os.path.isdir) can be true for the same path.

[Node.js](https://nodejs.org/docs/latest/api/fs.html#fslstatpath-options-callback):

> [fs.stat()](https://nodejs.org/docs/latest/api/fs.html#fsstatpath-options-callback) follows symbolic links. Use [fs.lstat()](https://nodejs.org/docs/latest/api/fs.html#fslstatpath-options-callback) to look at the links themselves.

[PHP](https://www.php.net/manual/en/function.is-dir.php):

> is_dir()
> It will return true if the file is actually a symlink that points to a directory.

Reply:
> Could we consider fixing the FShelpers in a separate PR so in the meantime we have a way to deal with symlinks? I'd be happy to help out with that.

That sounds reasonable to me 👍

Reply:
> At the FSHelpers level, it might make sense to follow symlinks. I've checked a few languages and it seems like their standard library follows symlinks for utilities like isFile and isDir

Thanks for looking into that! Given your findings, i looked a bit more and found that we can actually very easily support this behavior in `isFile` and `isDir` by passing the `follow` option:

```
FS.lookupPath('/path', { follow: true } );
```

Luckily, this also works great as a way to implement the realpath helper. While we don't necessarily need the helper methods I added in the PR to solve the issue, I think they still add value in general.

One use case is that in Studio, we need to mount symlink targets so being able to get readlink info via the PHP instance adds value for us.

I made some changes to rely on `lookupPath`

Reply:
Oh this updated PR looks lovely! I'm quite happy with it and would only ask to add a few unit tests for the new FSHelper methods. Other than that, it looks great.

Reply:
> Given your findings, i looked a bit more and found that we can actually very easily support this behavior in isFile and isDir by passing the follow option

This is great!

Reply:
> Oh this updated PR looks lovely! I'm quite happy with it and would only ask to add a few unit tests for the new FSHelper methods. Other than that, it looks great.

@jeroenpf, thanks for all your work on this! I'm adding some tests and plan to merge this yet today.